### DISCUSSION TITLE
How to unit test one struct method while mocking other methods?

### DISCUSSION THREAD

**Discussion by rharish101** [asomers/mockall#466] [Questions] [answered]
Say I have a struct `Foo` with two methods `method_1` and `method_2` such that `method_2` calls `method_1`:
```rust
struct Foo { ... };
struct Bar { ... };

impl Foo {
    fn method_1(&self) -> Bar {
        // Complex code
    }
    fn method_2(&self) {
        // Complex code
        let result = self.method_1();
        // Complex code
    }
}
```

How can I write a unit test for `method_2` by mocking just `method_1`? I tried adding `#[cfg(test, automock)]` before `method_1`, but I got: `message: automock does not support this item type`.

**Comment by asomers** [comment, marked as answer]
The short answer is that you can't.  Mockall replaces the entire object with a mock version.  You might be able to get away with the following, _if_ `method_2` doesn't access any of the struct's data:
```rust
#[automock]
impl Foo {
    fn method_1(&self) -> Bar {
        ...
    }
}
impl Foo {
    fn method_2(&self) {
        ....
    }
}
```
However, that strategy will break down if `method_2` tries to access any struct data.  Instead, a better strategy is to separate the object into two.  Not just in your test code, but in your product code as well.  Arrange it in layers, such that you can mock the lower layer when testing the higher one.  That's what I do.  Like this:
```rust
mod foo {
    struct Foo {...}
    #[automock]
    impl Foo {...}
}
mod bar {
    #[double]
    use crate::foo::Foo;

    struct Bar{foo: Foo, ...}
    #[automock]
    impl Bar {...}
}
mod baz {
    #[double]
    use crate::bar::Baz;

    struct Baz {bar: Bar, ....}
}
```

**Reply by rharish101** [reply]
Thanks a lot! I went with the second method, since I needed to mock `method_2` as well elsewhere in the codebase.
